src/libs/merge/kdbmerge.c

```diff
@@ -0,0 +1,527 @@
+			 *  This place is hit twice, thus overlap1empty gets double the amount of errors.
+			 */
+			if (keyInFirst != NULL)
+			{
+				existingKey = keyInFirst;
+			}
+			else if (keyInSecond != NULL)
+			{
+				existingKey = keyInSecond;
+			}
+			else
+			{
+				fprintf (stderr, "Something went wrong\n");
+				return -1;
+			}
+			if (thisConflict)
+			{
+				nonOverlapBaseEmptyCounter++;
+			}
+			if (!keysAreEqual (checkedKey, existingKey))
+			{
+				// overlap  with single empty
+				// This spot is hit twice for a single overlap conflict. Thus calculate half later on.
+				overlap1empty++;
+				if (checkedIsDominant) // TODO This also happens when there is no conflict
+				{
+					if (ksAppendKey (result, checkedKey) < 0)
+					{
+						fprintf (stderr, "Could not append key\n");
+						return -1;
+					}
+				}
+			}
+			else
+			{
+				// uses the NULL properties of keysAreEqual
+				if (keysAreEqual (checkedKey, keyInFirst) && baseIndicator == 2)
+				{
+					thisConflict = true;
+				}
+				if (keysAreEqual (checkedKey, keyInSecond) && baseIndicator == 1)
+				{
+					thisConflict = true;
+				}
+				if (thisConflict)
+				{
+					// base is empty and other and their have the same (non-empty) value
+					// this is a conflict
+					nonOverlapBaseEmptyCounter++;
+					if (checkedIsDominant)
+					{
+						if (ksAppendKey (result, checkedKey) < 0)
+						{
+							return -1;
+						}
+					}
+				}
+			}
+		}
+	}
+	return 0;
+}
+
+
+/**
+ * Returns merged key set
+ */
+KeySet * kdbMerge (KeySet * our, Key * ourRoot, KeySet * their, Key * theirRoot, KeySet * base, Key * baseRoot, Key * resultRoot,
+		   int strategy)
+{
+	ELEKTRA_LOG ("cmerge starts");
+	fprintf (stdout, "cmerge starts with strategy %d\n", strategy);
+	nonOverlapOnlyBaseCounter = 0;
+	nonOverlapBaseEmptyCounter = 0;
+	nonOverlapAllExistCounter = 0;
+	overlap3different = 0;
+	overlap1empty = 0;
+	KeySet * result = ksNew (0, KS_END);
+	KeySet * ourCropped = removeRoots (our, ourRoot);
+	KeySet * theirCropped = removeRoots (their, theirRoot);
+	KeySet * baseCropped = removeRoots (base, baseRoot);
+	ksRewind (ourCropped);
+	ksRewind (theirCropped);
+	ksRewind (baseCropped);
+	bool baseDominant = false;
+	bool ourDominant = false;
+	bool theirDominant = false;
+	switch (strategy)
+	{
+	case 3:
+		ourDominant = true;
+		break;
+	case 4:
+		theirDominant = true;
+		break;
+	case 5:
+		baseDominant = true;
+		break;
+	}
+
+	checkSingleSet (baseCropped, ourCropped, theirCropped, result, baseDominant, 0);
+	checkSingleSet (theirCropped, baseCropped, ourCropped, result, theirDominant, 1);
+	checkSingleSet (ourCropped, theirCropped, baseCropped, result, ourDominant, 2);
+	if (ksDel (ourCropped) != 0 || ksDel (theirCropped) != 0 || ksDel (baseCropped) != 0)
+	{
+		fprintf (stderr, "Could not delete keysets\n");
+	}
+	if (getTotalConflicts () > 0)
+	{
+		fprintf (stdout, "Conflict statistic:\n  %2d overlaps\n  %2d non-overlaps\n   --------------\n  %2d total\n",
```

Comment:
Statistics might be interesting for library users. You can return them in meta-data next to the errors/warnings.

Reply:
> next to the errors/warnings.

That means in an error key?

Reply:
yes